File: crates/mach/src/public/plugins/resolver/resolver.rs
use std::path::PathBuf;

use dependency::Dependency;

use crate::public::dependency;

pub trait Resolver {
    fn resolve(
      &self,
      dependency: &Dependency
    ) -> Result<Option<ResolveResult>, String>;
}

pub struct ResolveResult {
  pub file_path: PathBuf
}

File: npm/node-adapter/src/public/resolver.ts
export type Dependency = {
  id: string,
  specifier: string,
  specifier_type: string,
  is_entry: boolean,
  source_asset_id: string,
  source_path: string,
  resolve_from: string,
  imported_symbols: Array<string>,
}

export type ResolveOptions = {
  dependency: Dependency
}

export type ResolveResult = {
  file_path: string
}

export type ResolverInit = {
  resolve(options: ResolveOptions): (ResolveResult | undefined) | Promise<(ResolveResult | undefined)>
}

export class Resolver {
  init: ResolverInit

  constructor(init: ResolverInit) {
    this.init = init
  }
}
